### PR TITLE
CLI caching

### LINKED ISSUE #1300: Proposal: download Cypress to the common cached folder
## Feature

Instead of downloading and unzipping Cypress into `./node_modules/cypress/dist/` download it into a global folder for better caching. This will allow

- better caching of dependencies on CI
- reusing same version of Cypress if already downloaded once for some project

To determine output location use https://github.com/LinusU/node-cachedir

```js
var cachedir = require('cachedir')
var path = cachedir('cypress')
// `path` now contains the path under which you should store cached files
```

- cypress run, open, verify
  - [ ] ~take arg `--binary-folder` (looks like: `/foo/bar/Cypress.app` (macOS) or `/foo/bar/Cypress` (linux, win), to override default path to cypress binary folder (default is  `(cypressCache)/<cli version>/Cypress.app`)~
  - [x] can override folder of binary with env var `CYPRESS_BINARY_FOLDER=/path/to/Cypress.app`
    - [x] warn user if set

- cypress install
  - [x] Checks `(cypressCache)/<cli version>` for already installed binary folder (`Cypress.app` on macOS, `Cypress` on linux/win), and if binary version matches `(installVersion)`, do nothing.
  - [x] `(installVersion)` defaults to cli version, is overridden by env var `CYPRESS_BINARY_VERSION` during install
    - can be either `semvar: 1.2.3`, `url: www.cypress.io/download/1.2.3`, or `zip: /path/to/cypress.zip`
  - [x] if `cypress install --force`, force reinstall to `(cypressCache)/<cli version>`
  - [x] `(cypressCache)` defaults to:
    - `~/Library/Caches/Cypress` (macOS)
    - `~/.cache/Cypress` (linux)
    - `/AppData/Local/Cypress/Cache` (win)
  -  [x] `(cypressCache)` can be overridden with env var `CYPRESS_CACHE_DIRECTORY`
  - [x] before install, ensure write access to `(cypressCache)`

- [x] custom error message for failed access to `(cypressCache)`
- [ ] docs
  - [x] Installing Cypress
  - [ ] ~command line usage~
  - [ ] update info about global install
- [x] manual testing all use cases
  
  - [x] cypress install
    - CYPRESS_BINARY_VERSION
      - [x] url 
        - [x] skip if has any cache
      - [x] semvar
      - [x] zip
        - [x] skip if has any cache
        - [x] show progress for unzipping 
      - [x] --force
    - [x] --force
  - [x] ~--binary-folder~
    - [x] ~cypress run~
    - [x] ~cypress open~
    - [x] ~cypress verify~
    - [x] ~invalid~
  - [x] CYPRESS_CACHE_DIRECTORY
    - valid
    - invalid, error message
  

- [x] add local code coverage with `nyc`
  - [ ] ~~send to `coveralls` in CI~~
## Stretch goal
- [ ] can we cache `ffmpeg` separately to avoid bundling the same version over and over?

### PR DESCRIPTION
**Pull request by Bkucera**
closes #1300 #1254 #890

### CODE REVIEW

**Comment by Bkucera** [comment]
Code coverage with `nyc` snuck in here too

File: cli/__snapshots__/install_spec.js [hunk 1]
```diff
@@ -32,14 +31,22 @@ https://on.cypress.io/installing-cypress
 
 `
 
-exports['version already installed 1'] = `
-Cypress 1.2.3 is already installed. Skipping installation.
+exports['continues installing on failure 1'] = `
+Installed version ([object Object]) does not match needed version (1.2.3).
```
Comment (bahmutov) [inline cli/__snapshots__/install_spec.js]: JavaScript classic `[object Object]`

File: cli/lib/cli.js [hunk 2]
```diff
@@ -114,6 +116,7 @@ module.exports = {
     .option('--group', text('group'), coerceFalse)
     .option('--group-id <group-id>', text('groupId'))
     .option('--dev', text('dev'), coerceFalse)
+    .option('--cypress-path <cypress path>', text('cypressPath'))
```
Comment (bahmutov) [inline cli/lib/cli.js]: probably want to say this is `cypress folder` otherwise it might seem like Cypress app filepath

File: cli/lib/cli.js [hunk 3]
```diff
@@ -142,19 +147,23 @@ module.exports = {
...
     program
     .command('verify')
+    .usage('[options]')
     .description('Verifies that Cypress is installed correctly and executable')
-    .action(() => {
+    .option('--cypress-path <cypress path>', text('cypressPath'))
+    .action((opts) => {
       require('./tasks/verify')
-      .start({ force: true, welcomeMessage: false })
+      .start(_.extend(parseOpts(opts), { force: true, welcomeMessage: false }))
```
Comment (bahmutov) [inline cli/lib/cli.js]: could be better split into options + start
```
const defaultOption = { force: true, welcomeMessage: false }
const parsed = parseOpts(opts)
const options = _.extend(parsed, defaultOptions)
require('./tasks/verify')
  .start(options)
```

File: cli/package.json [hunk 4]
```diff
@@ -12,17 +12,24 @@
   "scripts": {
     "test": "npm run test-unit",
     "test-unit": "npm run dtslint && npm run unit",
-    "test-watch": "bin-up mocha --watch",
+    "test-watch": "npm run unit -- --watch",
```
Comment (bahmutov) [inline cli/package.json]: I get what you are doing - reusing `npm run unit` command, good

**Comment by bahmutov** [comment]
nice, nice, what are we doing to store `nyc` coverage?

also, can we think how we are going to test it - maybe on Circle after building binary... It can be done in a separate commits later after this lands.

File: cli/lib/cli.js (hunk 2 above)
Comment (Bkucera) [reply cli/lib/cli.js]: should we do `--cypress-dir` or `--cypress-folder`? or something longer like `--path-to-binary`

File: cli/package.json [hunk 5]
```diff
@@ -32,13 +39,15 @@
     "@types/bluebird": "3.5.18",
     "@types/chai": "4.0.8",
     "@types/chai-jquery": "1.1.35",
+    "@types/chalk": "2.2.0",
```
Comment (Bkucera) [inline cli/package.json]: why are `@types` in dependencies over devDependencies?

**Comment by Bkucera** [comment]
- [ ] Log to user where cypress binary was installed

**Comment by Bkucera** [comment]
changing binary directory to be `cache/Cypress/1.2.3/Cypress.app` (MacOS) instead of `cache/Cypress/1.2.3`, so users can specify the 'Cypress.app' folder as the binary instead of the parent of that folder
- [ ]

**Comment by brian-mann** [comment]
@Bkucera can you sign the CLA? Thanks

**Comment by jennifer-shehane** [comment]
@brian-mann I'm not sure what's happened with the cla-assistant on this PR, but it appears to be bugging. 

I manually checked and verified that @Bkucera signed the CLA for `cypress-io/cypress` on Apr 13, 2018 5:37:08 PM

**Comment by brian-mann** [comment]
I told the CLA assistant to recheck PR's and it's ready to be merged now.

@Bkucera did you have anything else to add to this PR or is it ready to go? Did you manually test everything, setting env vars, changing cache location, installing current version of Cypress, etc?

Did you try manually testing it from another project after Cypress has been installed from a different project?

**Comment by Bkucera** [comment]
@brian-mann trying to keep #1300 current with progress, but I need to finish
manual testing, custom error message for failed access to cache directory, and docs

**Pull request merged**